### PR TITLE
Smarter deserialize_bytes and deserialize_byte_buf

### PR DESCRIPTION
**Pull request by tikue**
It seems we can just read all at once rather than reading byte by byte.

### CODE REVIEW

**Comment by TyOverby** [comment]
This looks really good, thanks!  While you are in the read_vec code, could you make a little change?

```rust
let mut buffer = Vec::new();
try!(self.reader.by_ref().take(len as u64).read_to_end(&mut buffer));
```

could be replaced with
```
let mut bytes = Vec::with_capacity(len);
unsafe { bytes.set_len(len); }
try!(self.reader.read_exact(&mut bytes[..]));
```

and you'd probably see even better performance.

(this implementation was ruthlessly stolen from https://github.com/TyOverby/bincode/pull/116/files)

File: src/serde/reader.rs [hunk 1]
```diff
@@ -192,13 +195,13 @@ impl<'a, R: Read> serde::Deserializer for &'a mut Deserializer<R> {
     fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value>
         where V: serde::de::Visitor,
     {
-        self.deserialize_seq(visitor)
+        visitor.visit_byte_buf(try!(self.read_vec()))
```
Comment (TyOverby) [inline src/serde/reader.rs]: shouldn't this be a call to `visit_bytes`?

Comment (tikue) [inline src/serde/reader.rs]: Done.

**Comment by tikue** [comment]
Great suggestion, thanks -- done!

**Comment by dtolnay** [comment]
Still not as smart as it could be. I filed #118 to follow up in another PR.

**Comment by TyOverby** [comment]
@dtolnay: sounds good!  I'll merge this and start in on the recommendations in #118

**Pull request merged**